server/sublist_cache.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016-2023 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+type Cache struct {
+	data     [slCacheMax]atomic.Pointer[cacheEntry]
+	hashFunc func(string) int
+}
+
+type cacheEntry struct {
+	key   string
+	value *SublistResult
+}
+
+type stringStruct struct {
+	str unsafe.Pointer
+	len int
+}
+
+//go:noescape
+//go:linkname memhash runtime.memhash
```

Comment:
While the idea of the PR is interesting and the results promising, relying on `go:linkname` in this way is dangerous as it ties us into the behaviour of a specific Go compiler implementation. This probably means that NATS Server would no longer compile using `gccgo` for example.

Reply:
I changed it to use the maphash package (what is already used in the server).
The performance seems to be even a little bit better.

Sidenode:

the maphash package does the same here: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/hash/maphash/maphash_runtime.go

Reply:
Thanks for switching to the `maphash` package, indeed it does call out with `go:linkname` in the happy case, but more importantly, it has a `purego` fallback.